File: backend/services/cache/research_cache.py
"""
Research Cache Service

Provides intelligent caching for Google grounded research results to reduce API costs.
Only returns cached results for exact keyword matches to ensure accuracy.
"""

import hashlib
import json
from typing import Dict, Any, Optional, List
from datetime import datetime, timedelta
from loguru import logger


class ResearchCache:
    """Cache for research results with exact keyword matching."""
    
    def __init__(self, max_cache_size: int = 100, cache_ttl_hours: int = 24):
        """
        Initialize the research cache.
        
        Args:
            max_cache_size: Maximum number of cached entries
            cache_ttl_hours: Time-to-live for cache entries in hours
        """
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.max_cache_size = max_cache_size
        self.cache_ttl = timedelta(hours=cache_ttl_hours)
    
    def _generate_cache_key(self, keywords: List[str], industry: str, target_audience: str) -> str:
        """
        Generate a cache key based on exact keyword match.
        
        Args:
            keywords: List of research keywords
            industry: Industry context
            target_audience: Target audience context
            
        Returns:
            MD5 hash of the normalized parameters
        """
        # Normalize and sort keywords for consistent hashing
        normalized_keywords = sorted([kw.lower().strip() for kw in keywords])
        normalized_industry = industry.lower().strip() if industry else "general"
        normalized_audience = target_audience.lower().strip() if target_audience else "general"
        
        # Create a consistent string representation
        cache_string = f"{normalized_keywords}|{normalized_industry}|{normalized_audience}"
        
        # Generate MD5 hash
        return hashlib.md5(cache_string.encode('utf-8')).hexdigest()
    
    def _is_cache_entry_valid(self, entry: Dict[str, Any]) -> bool:
        """Check if a cache entry is still valid (not expired)."""
        if 'created_at' not in entry:
            return False
        
        created_at = datetime.fromisoformat(entry['created_at'])
        return datetime.now() - created_at < self.cache_ttl
    
    def _cleanup_expired_entries(self):
        """Remove expired cache entries."""
        expired_keys = []
        for key, entry in self.cache.items():
            if not self._is_cache_entry_valid(entry):
                expired_keys.append(key)
        
        for key in expired_keys:
            del self.cache[key]
            logger.debug(f"Removed expired cache entry: {key}")
    
    def _evict_oldest_entries(self, num_to_evict: int):
        """Evict the oldest cache entries when cache is full."""
        # Sort by creation time and remove oldest entries
        sorted_entries = sorted(
            self.cache.items(),
            key=lambda x: x[1].get('created_at', ''),
            reverse=False
        )
        
        for i in range(min(num_to_evict, len(sorted_entries))):
            key = sorted_entries[i][0]
            del self.cache[key]
            logger.debug(f"Evicted oldest cache entry: {key}")
    
    def get_cached_result(self, keywords: List[str], industry: str, target_audience: str) -> Optional[Dict[str, Any]]:
        """
        Get cached research result for exact keyword match.
        
        Args:
            keywords: List of research keywords
            industry: Industry context
            target_audience: Target audience context
            
        Returns:
            Cached research result if found and valid, None otherwise
        """
        cache_key = self._generate_cache_key(keywords, industry, target_audience)
        
        if cache_key not in self.cache:
            logger.debug(f"Cache miss for keywords: {keywords}")
            return None
        
        entry = self.cache[cache_key]
        
        # Check if entry is still valid
        if not self._is_cache_entry_valid(entry):
            del self.cache[cache_key]
            logger.debug(f"Cache entry expired for keywords: {keywords}")
            return None
        
        logger.info(f"Cache hit for keywords: {keywords} (saved API call)")
        return entry.get('result')
    
    def cache_result(self, keywords: List[str], industry: str, target_audience: str, result: Dict[str, Any]):
        """
        Cache a research result.
        
        Args:
            keywords: List of research keywords
            industry: Industry context
            target_audience: Target audience context
            result: Research result to cache
        """
        cache_key = self._generate_cache_key(keywords, industry, target_audience)
        
        # Cleanup expired entries first
        self._cleanup_expired_entries()
        
        # Check if cache is full and evict if necessary
        if len(self.cache) >= self.max_cache_size:
            num_to_evict = len(self.cache) - self.max_cache_size + 1
            self._evict_oldest_entries(num_to_evict)
        
        # Store the result
        self.cache[cache_key] = {
            'result': result,
            'created_at': datetime.now().isoformat(),
            'keywords': keywords,
            'industry': industry,
            'target_audience': target_audience
        }
        
        logger.info(f"Cached research result for keywords: {keywords}")
    
    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics."""
        self._cleanup_expired_entries()
        
        return {
            'total_entries': len(self.cache),
            'max_size': self.max_cache_size,
            'ttl_hours': self.cache_ttl.total_seconds() / 3600,
            'entries': [
                {
                    'keywords': entry['keywords'],
                    'industry': entry['industry'],
                    'target_audience': entry['target_audience'],
                    'created_at': entry['created_at']
                }
                for entry in self.cache.values()
            ]
        }
    
    def clear_cache(self):
        """Clear all cached entries."""
        self.cache.clear()
        logger.info("Research cache cleared")


# Global cache instance
research_cache = ResearchCache()

File: frontend/src/services/researchCache.ts
/**
 * Frontend Research Cache Service
 * 
 * Provides persistent caching for research results to survive page refreshes,
 * browser restarts, and internet disconnections.
 */

import { BlogResearchResponse } from './blogWriterApi';

interface CachedResearchEntry {
  result: BlogResearchResponse;
  keywords: string[];
  industry: string;
  target_audience: string;
  created_at: string;
  expires_at: string;
  cache_key: string;
}

interface CacheStats {
  total_entries: number;
  expired_entries: number;
  valid_entries: number;
  cache_size_mb: number;
}

class ResearchCacheService {
  private readonly CACHE_KEY_PREFIX = 'alwrity_research_';
  private readonly CACHE_VERSION = '1.0';
  private readonly DEFAULT_TTL_HOURS = 24;
  private readonly MAX_CACHE_SIZE_MB = 50; // 50MB max cache size

  /**
   * Generate a cache key for research parameters
   */
  private generateCacheKey(keywords: string[], industry: string, target_audience: string): string {
    const normalized_keywords = keywords.map(k => k.toLowerCase().trim()).sort();
    const normalized_industry = (industry || 'general').toLowerCase().trim();
    const normalized_audience = (target_audience || 'general').toLowerCase().trim();
    
    const cache_string = `${normalized_keywords.join(',')}|${normalized_industry}|${normalized_audience}`;
    const hash = this.simpleHash(cache_string);
    
    return `${this.CACHE_KEY_PREFIX}${this.CACHE_VERSION}_${hash}`;
  }

  /**
   * Simple hash function for cache keys
   */
  private simpleHash(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(36);
  }

  /**
   * Check if a cache entry is still valid
   */
  private isEntryValid(entry: CachedResearchEntry): boolean {
    const now = new Date();
    const expiresAt = new Date(entry.expires_at);
    return now < expiresAt;
  }

  /**
   * Clean up expired entries
   */
  private cleanupExpiredEntries(): void {
    const keys = Object.keys(localStorage);
    const expiredKeys: string[] = [];

    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        try {
          const entry = JSON.parse(localStorage.getItem(key) || '{}') as CachedResearchEntry;
          if (!this.isEntryValid(entry)) {
            expiredKeys.push(key);
          }
        } catch (error) {
          // Invalid JSON, remove it
          expiredKeys.push(key);
        }
      }
    });

    expiredKeys.forEach(key => {
      localStorage.removeItem(key);
      console.log(`Removed expired cache entry: ${key}`);
    });
  }

  /**
   * Get cache size in MB
   */
  private getCacheSizeMB(): number {
    let totalSize = 0;
    const keys = Object.keys(localStorage);
    
    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        const value = localStorage.getItem(key) || '';
        totalSize += key.length + value.length;
      }
    });

    return totalSize / (1024 * 1024); // Convert to MB
  }

  /**
   * Evict oldest entries if cache is too large
   */
  private evictOldestEntries(): void {
    const keys = Object.keys(localStorage);
    const cacheEntries: Array<{ key: string; created_at: string }> = [];

    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        try {
          const entry = JSON.parse(localStorage.getItem(key) || '{}') as CachedResearchEntry;
          cacheEntries.push({
            key,
            created_at: entry.created_at
          });
        } catch (error) {
          // Invalid entry, remove it
          localStorage.removeItem(key);
        }
      }
    });

    // Sort by creation time (oldest first)
    cacheEntries.sort((a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime());

    // Remove oldest entries until we're under the limit
    const maxSizeMB = this.MAX_CACHE_SIZE_MB;
    while (this.getCacheSizeMB() > maxSizeMB && cacheEntries.length > 0) {
      const oldest = cacheEntries.shift();
      if (oldest) {
        localStorage.removeItem(oldest.key);
        console.log(`Evicted oldest cache entry: ${oldest.key}`);
      }
    }
  }

  /**
   * Get cached research result
   */
  getCachedResult(keywords: string[], industry: string, target_audience: string): BlogResearchResponse | null {
    try {
      this.cleanupExpiredEntries();
      
      const cacheKey = this.generateCacheKey(keywords, industry, target_audience);
      const cachedData = localStorage.getItem(cacheKey);
      
      if (!cachedData) {
        console.log(`Cache miss for keywords: ${keywords.join(', ')}`);
        return null;
      }

      const entry = JSON.parse(cachedData) as CachedResearchEntry;
      
      if (!this.isEntryValid(entry)) {
        localStorage.removeItem(cacheKey);
        console.log(`Cache entry expired for keywords: ${keywords.join(', ')}`);
        return null;
      }

      console.log(`Cache hit for keywords: ${keywords.join(', ')} (saved API call)`);
      return entry.result;
    } catch (error) {
      console.error('Error retrieving cached result:', error);
      return null;
    }
  }

  /**
   * Cache a research result
   */
  cacheResult(
    keywords: string[], 
    industry: string, 
    target_audience: string, 
    result: BlogResearchResponse,
    ttlHours: number = this.DEFAULT_TTL_HOURS
  ): void {
    try {
      this.cleanupExpiredEntries();
      this.evictOldestEntries();

      const cacheKey = this.generateCacheKey(keywords, industry, target_audience);
      const now = new Date();
      const expiresAt = new Date(now.getTime() + (ttlHours * 60 * 60 * 1000));

      const entry: CachedResearchEntry = {
        result,
        keywords,
        industry,
        target_audience,
        created_at: now.toISOString(),
        expires_at: expiresAt.toISOString(),
        cache_key: cacheKey
      };

      localStorage.setItem(cacheKey, JSON.stringify(entry));
      console.log(`Cached research result for keywords: ${keywords.join(', ')}`);
    } catch (error) {
      console.error('Error caching result:', error);
    }
  }

  /**
   * Get cache statistics
   */
  getCacheStats(): CacheStats {
    this.cleanupExpiredEntries();
    
    const keys = Object.keys(localStorage);
    let totalEntries = 0;
    let expiredEntries = 0;
    let validEntries = 0;

    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        totalEntries++;
        try {
          const entry = JSON.parse(localStorage.getItem(key) || '{}') as CachedResearchEntry;
          if (this.isEntryValid(entry)) {
            validEntries++;
          } else {
            expiredEntries++;
          }
        } catch (error) {
          expiredEntries++;
        }
      }
    });

    return {
      total_entries: totalEntries,
      expired_entries: expiredEntries,
      valid_entries: validEntries,
      cache_size_mb: this.getCacheSizeMB()
    };
  }

  /**
   * Clear all cached research results
   */
  clearCache(): void {
    const keys = Object.keys(localStorage);
    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        localStorage.removeItem(key);
      }
    });
    console.log('Research cache cleared');
  }

  /**
   * Get all cached entries (for debugging)
   */
  getAllCachedEntries(): CachedResearchEntry[] {
    const entries: CachedResearchEntry[] = [];
    const keys = Object.keys(localStorage);

    keys.forEach(key => {
      if (key.startsWith(this.CACHE_KEY_PREFIX)) {
        try {
          const entry = JSON.parse(localStorage.getItem(key) || '{}') as CachedResearchEntry;
          entries.push(entry);
        } catch (error) {
          console.error(`Error parsing cache entry ${key}:`, error);
        }
      }
    });

    return entries.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
  }
}

// Export singleton instance
export const researchCache = new ResearchCacheService();
export default researchCache;
